python/ray/data/_internal/execution/interfaces/physical_operator.py

```diff
@@ -276,6 +276,12 @@ def mark_execution_completed(self):
         """Manually mark this operator has completed execution."""
         self._execution_completed = True
 
+    def execution_completed(self) -> bool:
+        """Return True when this operator has completed execution.
+        The outputs may or may not have been taken.
+        """
+        return self._execution_completed
+
```

Comment:
I'd suggest to fuse this method into completed (just adding default param to it):

 - This is a sub-scope of it 
 - It's confusing to have 2 separate method w/o clear distinction

Reply:
You mean something like `completed(execution_completed=True)`?
I find the current way more clear

Reply:
I also felt confused reasoning about the difference between `completed` and `execution_completed`. 

Here’s how I currently understand it:
- We use `completed` to decide:
  - Whether to autoscale an actor pool (though we *could* use `execution_completed` instead)
  - Whether an operator should keep receiving inputs
- We use `execution_completed` to decide:
  - Whether an operator is eligible for resource reservation
  - Whether the operator is `completed`

Since we use them in different ways, I’m not sure how we’d combine them into one method. But maybe this would help:
	1.	Rename `execution_completed` to something like `is_done_launching_tasks`
	2.	Use `is_done_launching_tasks` for autoscaling instead of `completed`

Reply:
@raulchen they couldn't have `completed` and `execution_completed` names.

My proposal is to either

 - `completed(finalized=True)`
 - Keep 2 methods renaming to `execution_finished` and `finalized`

Reply:
what about remaining `execution_completed` to `execution_stopped`? 
This API basically means all tasks have finished and not more tasks will be launched. so `is_done_launching_tasks` is not accurate.
And the name also aligns with the doc-string of `completed`
```
        """Return True when this operator is completed.

        An operator is completed the operator has stopped execution and all
        outputs are taken.
```

Reply:
also agreed that autoscaling (as well as a few other places) should use `execution_completed` instead of `completed`. 
I'll handle that in a follow-up PR

Reply:
actually I just realized that it doesn't make a big difference to distinguish between these 2 cases. 
Because `execution_completed` will [become `completed` immediately](https://github.com/ray-project/ray/blob/5d9e66150e0ad34cbb4d4d5376db78f61021ce93/python/ray/data/_internal/execution/streaming_executor_state.py#L501-L503).
I'll just revert this change to avoid make it overcomplicated